47_agent_ip_port.md

```diff
@@ -0,0 +1,182 @@
+- [ ] Test plan is in place
+- [ ] User-facing documentation has been created in [keylime/keylime-docs]
+
+<!--
+**Note:** This checklist is iterative and should be reviewed and updated every time this enhancement is being considered for a milestone.
+-->
+
+## Summary
+The agents gains the ability to provide the registrar with a contact ip and port
+for the verifier.
+
+## Motivation
+The verifier needs a port and ip adress to contact a agent. 
+Currently this information is manually provided to the tenant.
+
+This information can also be provided by the agent in most cases. It makes
+Keylime more user friendly in environments where no external configuration
+management is used.
+
+Allowing configuration with environment variables allows other tools to set this
+information easily before the agent is started.
+### Goals
+ * Add support for the agent to specify a contact ip and port
+  * Static configuration in config file or specified via environment variable
+ * Extending registrar API to allow specifying contact ip and port 
+ * Add support for using the specified ip and port to the tenant
+
+### Non-Goals
+ * Moving Keylime to model to a push only model
+ * Allow IP changes after registration
+ * Auto detection of IP and port
+
+## Proposal
+This enhancement proposal adds that the agent can provide a contact IP and port
+for the verifier to the registrar. The tenant does not have to ask the user
+manually for this data anymore and tries to retrieve it from the registrar. 
+
+### User Stories
+
+#### Story 1
+ * Agent has contact ip and port configured
+ * Agent registers itself with the registrar
+ * User adds agent with `keylime_tenant -c add -u AGENT_ID`
+ * Tenant retrieves the ip and port information from the registrar
+ * Agent is added to the verifier
+
+#### Story 2
+ * Agent has contact ip and port not configured
+ * Agent registers itself with the registrar
+ * User adds agent with `keylime_tenant -c add -t 127.0.0.1 -u AGENT_ID`
+   (`127.0.0.1` is the IP of the agent)
+ * Agent is added to the verifier
+
+#### Story 3
+ * Agent has contact ip and port not configured
+ * Agent registers itself with the registrar
+ * User adds agent with `keylime_tenant -c add -u AGENT_ID`
+ * Tenant fails because the IP cannot be retrieved from the registrar
+
+### Risks and Mitigations
+The input by the agent is generally not trusted and must be validated.
+Is is done in the registrar. 
+
+## Design Details
+To the registrar database table `registrarmain` two new columns are added: `ip`
+and `port`. Those entries can be NULL.
+
+The fields `ip` and `port` can be optionally specified in the registrar API when
+a agents tries to register. Simple input validation for those fields is added in
+the registrar.
+
+The agent configuration gains two new optional fields `cloudagent_contact_ip`
+and `cloudagent_contact_port`. Those options can also be specified as
+environment variables with `CLOUDAGENT_CONTACT_IP` and
```

Comment:
Env vars should probably be prefixed by `KEYLIME_` since they are global and should immediately be apparent which program is using them. And since that makes it pretty long, we can probably drop the `CLOUD` part, so it would be `KEYLIME_AGENT_CONTACT_IP` or similar.

Reply:
I'll change that. What is the general consensus of Keylime being configurable with environment variables and where should they be documented?  

Currently the configuration options in `keylime.conf` also use `cloud` as a prefix should that be also changed?

Reply:
Ideally, we should aim to have the ability to "override" any parameter on `keylime.conf` directly via an environment variable. We are not there yet (and maybe never will), but the point made by @mpeters is an important one: let's make sure that at least new functionality introduced follows this pattern.

Reply:
Ok if that is the case we could modify the config parsing such that every key in the config could be set with `KEYLIME_SECTION_KEY` as environment variable. The advantage of that is that the naming would be always consistent between `keylime.conf` and the environment. Such a change should be probably its own enhancement proposal.

Reply:
Definitely agree, it's especially useful when running in a container as it's easier to do the config via env vars vs generating a config in an entrypoint.